### PR TITLE
type-safe `getSession()`

### PR DESCRIPTION
**Pull request by KATT**
- As part of #473 I saw a lot of occurrences of an error using `getSession()`
- Seems like we're manually extending the jwt + session to include some more stuff
- Here I added a specific `CalendsoSession` that adheres to what is defined in `[...nextauth.ts]`

Possibly there's a nicer way to extend the session in the first place, but this fixes some of the typing issues.

### CODE REVIEW

**Review by PeerRich — APPROVED** [review]

File: pages/api/auth/[...nextauth].tsx [hunk 1]
```diff
@@ -1,59 +1,73 @@
...
 
-                if (!user) {
-                    throw new Error('No user found');
-                }
+        if (!user) {
+          throw new Error("No user found");
+        }
+        if (!user.password) {
+          throw new Error("Incorrect password");
+        }
```
Comment (KATT) [inline pages/api/auth/[...nextauth].tsx:34]: Had to add this to get rid of type error as it's nullable

File: lib/auth.ts [hunk 2]
```diff
@@ -1,11 +1,29 @@
...
 
-export async function verifyPassword(password, hashedPassword) {
-    const isValid = await compare(password, hashedPassword);
-    return isValid;
-}
\ No newline at end of file
+export async function verifyPassword(password: string, hashedPassword: string) {
+  const isValid = await compare(password, hashedPassword);
+  return isValid;
+}
```
Comment (KATT) [inline lib/auth.ts]: added types on these

File: lib/auth.ts [hunk 3]
```diff
@@ -1,11 +1,29 @@
...
+  username: string;
+};
+export interface CalendsoSession extends DefaultSession {
+  user?: CalendsoSessionUser;
+}
+
+export async function getSession(options: GetSessionOptions): Promise<CalendsoSession | null> {
+  const session = await getSessionInner(options);
+
+  // that these are equal are unsured in `[...nextauth]`'s callback
```
Comment (KATT) [inline lib/auth.ts:27]:
```suggestion
  // that these are equal are ensured in `[...nextauth]`'s callback
```

**Review by KATT — COMMENTED** [review]
Highlighted my changes

**Comment by emrysal** [comment]
Thinking, do we actually need the `Calendso` prefix? I struggle coming up with good reasons to namespace this expose. Also, would it make sense to expose `(Calendso)SessionUser` as well?

(If Session conflicts with another import called Session; even though unlikely, we can always alias it in the import. `import { Session as CalendsoSession } from "@lib/auth";

File: lib/auth.ts [hunk 4]
```diff
@@ -1,11 +1,29 @@
...
+  const isValid = await compare(password, hashedPassword);
+  return isValid;
+}
+
+type DefaultSessionUser = NonNullable<DefaultSession["user"]>;
+type CalendsoSessionUser = DefaultSessionUser & {
+  id: number;
+  username: string;
+};
+export interface CalendsoSession extends DefaultSession {
```
Comment (KATT) [inline lib/auth.ts:20]:
```suggestion
interface Session extends DefaultSession {
```

@emrysal you'd prefer this?

**Pull request merged**